include/real/real_explicit.hpp

```diff
@@ -221,102 +222,82 @@ namespace boost {
-                for (const auto& c : number ) {
-                    if (c == '.' && dot_amount >= 1) {
-                        throw boost::real::invalid_string_number_exception();
-                    } else if (c == '.') {
-                        dot_amount++;
-                    }
+                std::size_t dot_pos = number.find('.');
+                if (dot_pos == std::string::npos) {
+                    integer_part = number;
+                    decimal_part = "";
                 }
-
-                bool there_is_dot = dot_amount > 0;
-                int exponent = 0;
-                dot_amount = 0;
-
-                // Check whether the number is explicitly specified as positive or negative.
-                unsigned int first_index = 0;
-                if (number.at(first_index) == '+') {
+                else {
+                    integer_part = number.substr(0, dot_pos);
+                    if(has_exponent)
+                        decimal_part = number.substr(dot_pos+1, e_pos - dot_pos -1);
+                    else
+                        decimal_part = number.substr(dot_pos+1);
+                }
+                if(integer_part[0]=='+') {
                     this->_positive = true;
-                    first_index++;
-                } else if (number.at(first_index) == '-') {
+                    integer_part = integer_part.substr(1);
+                }
+                if(integer_part[0]=='-') {
                     this->_positive = false;
-                    first_index++;
+                    integer_part = integer_part.substr(1);
                 }
-
-                // Remove zeros from the lefts side
-                // Note: We know at this point that number.length > 0 because of the first check
-                unsigned int last_index = (unsigned int)number.length() - 1;
-                while (last_index > first_index && (number.at(last_index) == '0' || number.at(last_index) == '.')) {
-                    if (number.at(last_index) == '.') {
-                        dot_amount++;
-                    }
-                    last_index--;
-                    if (!there_is_dot || dot_amount > 0) {
-                        exponent++; // remove zeros from the integer part increases the exponent
+                int i = 0;
+                while (integer_part[i]=='0'&&i<integer_part.length()) {
+                    ++i;
+                }
+                integer_part = integer_part.substr(i);
+                i = decimal_part.length()-1;
+                while (decimal_part[i]=='0'&&i>0) {
+                    --i;
+                }
+                decimal_part = decimal_part.substr(0, i+1);
+                //decimal and integer parts stripped of zeroes
+                int exponent = integer_part.length() + add_exponent;
+                if(decimal_part.empty()) {
+                    i = integer_part.length()-1;
+                    while (integer_part[i]=='0'&&i>0)
```

Comment:
When you strip zeros from the right side of the integer part, the exponent should be increased. I think this is a bug

Reply:
I had initially done that as it seemed to be a bug to me as well. But then I noticed that the exponent is already calculated as integer_part.length(). So ++exponent not needed. Plus, if only Travis CI was still working, I could demonstrate that all the tests are passing. But for now, please just take my word for it :-P

Reply:
1) In the Test there is no test of the case I say 12300.00_r that is why your tests passes. If you add the test I suggested I'm sure it wouldn't pass.

2) The exponent calculation integer_part.length() is done in the line 273 and you strip zeros in line 276, then if you have 12300.00 first at line 273 it will calculate an exponent of 5 (integer_part = 12300) but later in line 276 you strip the two zeros from the right and the integer_part become 123 while the exponent remains 5 as you are not updating it.

You can fix the bug just by moving the exponent calculation some lines so it is calculated once the integer_part is fully procesed. 

By the way, you could strip zeros using a regex instead of iterating to be consistent in the implementation.